### ISSUE TITLE
Symmetric logarithmic plotting

### ISSUE THREAD

**Issue by Cazadorro** [epezent/implot#245]
For most data I deal with, I don't really care about displaying values lograrithmicaly between -1 -> 1, and in fact, this gets in the way of plotting. Other plotting APIs use the term "symmetric log plotting" to handle this, ie matplot lib:

https://matplotlib.org/stable/gallery/scales/symlog_demo.html

see here for [more details ](https://stackoverflow.com/questions/39988048/what-is-the-origin-of-matplotlibs-symlog-a-k-a-symmetrical-log-scale).

Basically symlog/symmetric log plots log but only after a certain absolute value threshold.  This allows it to plot around zero.  One benefit of this is that, for example, the example histogram plot does not appear to extend all the way down to infinity.   

As an example use case where this would be helpful, I'm trying to plot histograms that have some extreme value differences. When switching to a normal log plot, all the bars look the same and the Y axis space taken up becomes very very large (due to them all encompassing 1).   The graph is pretty much un-navigable and impossible to read.  With symmetric log plotting in matplotlib, the values become visible within one screen with out special axis grid spacing rules. 

Looking through the code it looks like a change like this may be difficult, I don't exactly understand how  flags propagate through ImPlot.  Additionally there would need to be some way to feed in this "Linear threshold" parameter to ImPlot.  But in the future there's likely going to be more plotting functions that should be used.   In those cases there's likely going to be a need to specify user defined functions for mapping to avoid work for every function needed. 

Matplotlib takes straight functions and is able to use those to plot.  ImPlot appears to rely on being able to transform between pixels and plot space and vice versa, and I think Matplotlib does as well (forward and inverse functions must be provided in matplotlib) To support custom functions currently then, two functions would have to be provided. Allowing user defined functions for plotting mitigates issues if the API doesn't handle certain edge cases. 

To that end it would be nice to have: 

`ImPlotAxisFlags_SymLogScale` which maintains the same exclusion as `ImPlotAxisFlags_LogScale`, and adds exclusion to both (so log can't be used along with symlog). 

Some method of setting the "linear threshold" parameter, either through some sort of style push/pop or something else

`ImPlotAxisFlags_CustomScale` which uses some user specified function to plotting function (forward/inverse) potentially with the signature `double(double x, void* user_data)`, or if C++11 features are fine to add: `std::function<double(double)>` where user data can be added as a closure parameter and not have to specifically be dealt with by the API.  Right now it is possible to create a custom plot, but for only scaling it is kind of overkill since all work for any graph you want to user the scaling for needs to be re-done.

**Comment by epezent** [comment]
@Cazadorro , thanks for the idea. This is neat! 

I also found this relevant post:
https://stackoverflow.com/questions/3305865/what-is-the-difference-between-log-and-symlog
A transformation formula:
https://iopscience.iop.org/article/10.1088/0957-0233/24/2/027001/meta
And implementation:
https://www.mathworks.com/matlabcentral/fileexchange/57902-symlog

My initial reaction is that a `ImPlotAxisFlags_SymLogScale` should be possible, and maybe even easy to implement. Let me dig a little deeper and report back. 

Regarding, `ImPlotAxisFlags_CustomScale` and custom scaling callbacks, ImPlot is going to require some internal refactoring and decoupling of systems to enable something like that. These sorts of "callback based customizations" have been requested in other areas, e.g. input handling. It's probably quite a ways out, unless someone besides me steps to the plate. User callbacks are also not very "ImGui-esque" for what that is worth, and any customization beyond what the library ships with is probably best done by modifying the source code anyway (after all, the library is quite small).

**Comment by epezent** [comment]
Basic support for symlog plots are now supported with `ImPlotScale_SymLog` (see #370). 

![image](https://user-images.githubusercontent.com/29577475/174466135-aa6674f5-47cb-47c2-ae3f-f5ece24a277d.png)

Internally, this is using the following forward and inverse scaling functions

```cpp
static inline double TransformForward_SymLog(double v, void*) {
    return 2.0 * ImAsinh(v / 2.0);
}

static inline double TransformInverse_SymLog(double v, void*) {
    return 2.0 * ImSinh(v / 2.0);
}
```

so that the linear region is roughly between -1 and 1, and exponentialy below and above that. We also now support user define scaling functions (but not tick locators, yet).
